lightning/src/routing/network_graph.rs

```diff
@@ -1085,6 +1085,63 @@ impl NetworkGraph {
 		self.update_channel_from_unsigned_announcement_intern(msg, None, chain_access)
 	}
 
+	/// Update channel from partial announcement data received via rapid gossip
+	///
+	/// `timestamp: 64`: Timestamp the announcement was originally received.
+	///
+	/// All other parameters as used in `UnsignedChannelAnnouncement` fields.
+	pub fn update_channel_from_announcement_fragments(&self, short_channel_id: u64, timestamp: u64, features: ChannelFeatures, node_id_1: PublicKey, node_id_2: PublicKey) -> Result<(), LightningError>{
+		if node_id_1 == node_id_2 {
+			return Err(LightningError{err: "Channel announcement node had a channel with itself".to_owned(), action: ErrorAction::IgnoreError});
+		};
+
+		let node_1 = NodeId::from_pubkey(&node_id_1);
+		let node_2 = NodeId::from_pubkey(&node_id_2);
+		let chan_info = ChannelInfo {
+			features: features.clone(),
+			node_one: node_1.clone(),
+			one_to_two: None,
+			node_two: node_2.clone(),
+			two_to_one: None,
+			capacity_sats: None,
+			announcement_message: None,
+			announcement_received_time: timestamp,
+		};
+
+		let mut channels = self.channels.write().unwrap();
+		let mut nodes = self.nodes.write().unwrap();
+		match channels.entry(short_channel_id) {
+			BtreeEntry::Occupied(_) => {
+				// given that duplicate announcements are fine with a gossip sync scenario,
+				// perhaps we could just return `Ok(())` here
+				return Err(LightningError{err: "Already have knowledge of channel".to_owned(), action: ErrorAction::IgnoreDuplicateGossip});
+			},
+			BtreeEntry::Vacant(entry) => {
+				entry.insert(chan_info);
+			}
+		};
+
+		let mut add_channel_to_node = | node_id: NodeId | {
+			match nodes.entry(node_id) {
```

Comment:
Can you DRY this with the below function?

Reply:
done, I think